rfcs/text/0014_api_documentation.md

```diff
@@ -0,0 +1,439 @@
+  FunctionKind = 'Function',
+  ObjectKind = 'Object',
+  InterfaceKind = 'Interface',
+  TypeKind = 'Type', // For things like `export type Foo = ...`
+  Unknown = 'Unknown', // There are a lot of ts-morph types, if I encounter something not handled, I dump it in here.
+  String = 'string',
+  Number = 'number',
+  Boolean = 'boolean',
+  Array = 'Array',
+  CompoundType = 'CompoundType', // Unions & intersections, to handle things like `string | number`. 
+}
+```
+
+
+### Text with reference links
+
+Signatures, descriptions and returns comments may all contain links to other API declarations. This information needs to be serializable into json. This serializable type encompasses the information needed to build the DocLink components within these fields. The logic of building
+the DocLink components currently resides inside the elasticdocs system. It's unclear if this will change.
+
+```ts
+/**
+ * This is used for displaying code or comments that may contain reference links. For example, a function
+ * signature that is `(a: import("src/plugin_b").Bar) => void` will be parsed into the following Array:
+ *
+ * ```ts
+ * [
+ *   '(a: ',
+ *   { docId: 'pluginB', section: 'Bar', text: 'Bar' },
+ *   ') => void'
+ * ]
+ * ```
+ *
+ * This is then used to render text with nested DocLinks so it looks like this:
+ *
+ * `(a: => <DocLink docId="pluginB" section="Bar" text="Bar"/>) => void`
+ */
+export type TextWithLinks = Array<string | Reference>;
+
+/**
+ * The information neccessary to build a DocLink.
+ */
+export interface Reference {
+  docId: string;
+  section: string;
+  text: string;
+}
+```
+
+### ScopeApi
+
+Scope API is essentially just grouping an array of ApiDeclarations into different categories that makes building the mdx files from a
+single json file easier.
+
+```ts
+export interface ScopeApi {
+  setup?: ApiDeclaration;
+  start?: ApiDeclaration;
+  classes: ApiDeclaration[];
+  functions: ApiDeclaration[];
+  interfaces: ApiDeclaration[];
+  objects: ApiDeclaration[];
+  enums: ApiDeclaration[];
+  misc: ApiDeclaration[];
+  // We may add more here as we sit fit to pull out of `misc`.
+}
+```
+
+With this structure, the mdx files end up looking like:
+
+```
+### Start
+<DocDefinitionList data={[actionsJson.server.start]}/>
+### Functions
+<DocDefinitionList data={actionsJson.server.functions}/>
+### Interfaces
+<DocDefinitionList data={actionsJson.server.interfaces}/>
+```
+
+### PluginApi
+
+A plugins API is the component that is serialized into the json file. It is broken into public, server and common components.  `serviceFolders` is a way for the system to
+write separate mdx files depending on where each declaration is defined. This is because certain plugins (and core)
+are huge, and can't be rendered in a single page.
+
+
+```ts
+export interface PluginApi {
+  id: string;
+  serviceFolders?: readonly string[];
+  client: ScopeApi;
+  server: ScopeApi;
+  common: ScopeApi;
+}
+```
+
+## Manifest file changes
```

Comment:
Do you think it'd add more clarity if we referred to this as the "`kibana.json` Manifest file", instead of just the "Manifest file"?

Reply:
Yes, good idea.